### PR TITLE
Asset versioning docs

### PR DESCRIPTION
**Pull request by smackesey**
### Summary & Motivation

Adds some docs for source asset observations and asset code versions.

### How I Tested These Changes

Added new `docs_snippets` tests.

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #10922** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10922" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/10922?utm_source=stack-comment).

File: docs/content/concepts/assets/asset-observations.mdx [hunk 1]
```diff
@@ -105,3 +105,28 @@ def partitioned_dataset_op(context):
     )
     return df
 ```
+
+### Observable source assets
+
+<PyObject object="SourceAsset" /> objects may have a user-defiend _observation function_
```
Comment (sryza) [inline docs/content/concepts/assets/asset-observations.mdx:111]:
```suggestion
<PyObject object="SourceAsset" /> objects may have a user-defined _observation function_
```

File: docs/content/concepts/assets/asset-observations.mdx [hunk 2]
```diff
@@ -105,3 +105,28 @@ def partitioned_dataset_op(context):
...
+that returns a <PyObject object="LogicalVersion" />. Whenever the observation
+function is run, an `AssetObservation` will be generated for the source asset
+and tagged with the returned logical version. The logical version is used in
+staleness calculations for downstream assets.
+
+The <PyObject object="observable_source_asset" /> decorator provides a conveninet way to define source assets with observation functions. The below observable source asset takes a file hash and returns it as the logical version. Every time you run the observation function, a new observation will be generated with this hash set as logical verison.
```
Comment (sryza) [inline docs/content/concepts/assets/asset-observations.mdx:117]:
```suggestion
The <PyObject object="observable_source_asset" /> decorator provides a convenient way to define source assets with observation functions. The below observable source asset takes a file hash and returns it as the logical version. Every time you run the observation function, a new observation will be generated with this hash set as logical version.
```

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 3]
```diff
@@ -260,6 +260,34 @@ def downstream(may_not_materialize):
     return may_not_materialize + [5]
 ```
 
+### Asset code versions
+
+Assets may be assigned a `code_version`:
```
Comment (sryza) [inline docs/content/concepts/assets/software-defined-assets.mdx:265]: I would give some motivation for why someone would do this: e.g. "Versions let you help Dagster track what assets are _stale_ and avoid performing redundant computation."

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 4]
```diff
@@ -260,6 +260,34 @@ def downstream(may_not_materialize):
...
+
+```python file=/concepts/assets/code_versions.py startafter=start_single_asset endbefore=end_single_asset
+@asset(code_version="1")
+def asset_with_version():
+  return 100
+```
+
+When an asset with a code version is materialized, the generated `AssetMaterialization` is tagged with the version. The code version together with the <PyObject object="LogicalVersion" /> values of upstream assets are hashed to create a `LogicalVersion` for the asset, also stored on the `AssetMaterialization`, which is used to track asset staleness. Updating the `code_version` will change the projected `LogicalVersion` of the asset for a new materialization, thus rendering it stale.
```
Comment (sryza) [inline docs/content/concepts/assets/software-defined-assets.mdx:273]: I think there's a way to explain this that doesn't require readers to understand the details of the implementation / hashing etc.  E.g. "An asset that has a different code version than the code version used for its most recent materialization will be considered _stale_. Any assets downstream of a stale asset are also considered stale.".

**Review by sryza — APPROVED** [review]

**Pull request merged**